opencensus/stats/internal/delta_producer.h

```diff
@@ -0,0 +1,121 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPENCENSUS_STATS_INTERNAL_DELTA_PRODUCER_H_
+#define OPENCENSUS_STATS_INTERNAL_DELTA_PRODUCER_H_
+
+#include <cstdint>
+#include <memory>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "absl/synchronization/mutex.h"
+#include "absl/time/time.h"
+#include "opencensus/stats/bucket_boundaries.h"
+#include "opencensus/stats/distribution.h"
+#include "opencensus/stats/internal/measure_data.h"
+#include "opencensus/stats/measure.h"
+#include "opencensus/stats/tag_set.h"
+
+namespace opencensus {
+namespace stats {
+
+// Delta is thread-compatible.
+class Delta final {
+ public:
+  void Record(std::initializer_list<Measurement> measurements, TagSet tags);
+
+  void clear();
+  void SwapAndReset(
+      std::vector<std::vector<BucketBoundaries>>& registered_boundaries,
+      Delta* other);
+
+  void Consume();
+
+ private:
+  // A copy of registered_boundaries_ in the DeltaProducer as of when the
+  // delta was started.
+  std::vector<std::vector<BucketBoundaries>> registered_boundaries_;
+
+  // The actual data. Each MeasureData[] contains one element for each
+  // registered measure.
+  std::unordered_map<TagSet, std::vector<MeasureData>, TagSet::Hash> delta_;
+};
+
+// DeltaProducer is thread-safe.
+class DeltaProducer final {
+ public:
+  // Returns a pointer to the singleton DeltaProducer.
+  static DeltaProducer* Get();
+
+  // Adds a new Measure.
+  void AddMeasure();
+
+  // Adds a new BucketBoundaries for the measure 'index' if it does not already
+  // exist.
+  void AddBoundaries(uint64_t index, const BucketBoundaries& boundaries);
+
+  void Record(std::initializer_list<Measurement> measurements, TagSet tags)
+      LOCKS_EXCLUDED(delta_mu_);
+
+  // Flushes the active delta and blocks until it is harvested.
+  void Flush() LOCKS_EXCLUDED(delta_mu_, harvester_mu_);
+
+ private:
+  DeltaProducer();
+
+  // Flushing has two stages: swapping active_delta_ to last_delta_ and
+  // consuming last_delta_. Callers should release delta_mu_ before calling
+  // ConsumeLastDelta so that Record() is blocked for as little time as
+  // possible.
+  void SwapDeltas() EXCLUSIVE_LOCKS_REQUIRED(delta_mu_, harvester_mu_);
+  void ConsumeLastDelta() EXCLUSIVE_LOCKS_REQUIRED(harvester_mu_)
+      LOCKS_EXCLUDED(delta_mu_);
+
+  // Loops calling FlushInternal() every harvest_interval_.
```

Comment:
There is no FlushInternal. Maybe outdated comment.

Reply:
Fixed.